symbolic-base/src/ZkFold/Symbolic/Data/List.hs

```diff
@@ -1,45 +1,75 @@
+import           Data.List.Infinite               (Infinite (..))
 import           Data.Proxy                       (Proxy (..))
-import           Data.Traversable                 (Traversable)
-import           Data.Zip                         (Zip)
-import           GHC.Generics                     (Par1 (..))
-import           Prelude                          (fmap, fst, pure, type (~), undefined, ($), (.), (<$>))
+import           Data.Traversable                 (Traversable, traverse)
+import           Data.Tuple                       (snd)
+import           GHC.Generics                     (Generic, Generic1, Par1 (..), (:*:) (..), (:.:) (..))
+import           Prelude                          (fmap, fst, type (~), undefined, ($), (.), (<$>))
 
 import           ZkFold.Base.Algebra.Basic.Class
-import           ZkFold.Base.Data.Utils           (zipWithM)
+import           ZkFold.Base.Control.HApplicative (HApplicative)
+import           ZkFold.Base.Data.Functor.Rep     ()
+import           ZkFold.Base.Data.HFunctor        (hmap)
+import           ZkFold.Base.Data.List.Infinite   ()
+import           ZkFold.Base.Data.Product         (fstP, sndP)
 import           ZkFold.Symbolic.Class
 import           ZkFold.Symbolic.Data.Bool        (Bool (..))
 import           ZkFold.Symbolic.Data.Class
 import           ZkFold.Symbolic.Data.Combinators
-import           ZkFold.Symbolic.Data.Conditional
+import           ZkFold.Symbolic.Data.Input       (SymbolicInput)
+import           ZkFold.Symbolic.Data.Payloaded   (Payloaded (Payloaded))
 import           ZkFold.Symbolic.MonadCircuit
 
-data List (context :: (Type -> Type) -> Type) x
-    = List
-        { lHash    :: context (Layout x)
-        , lSize    :: context Par1
-        , lWitness :: [x]
-        -- ^ TODO: As the name suggests, this is only needed in witness cinstruction in uncons.
-        -- This list is never used in circuit itlest.
-        -- Think of a better solution for carrying witnesses.
-        -- Besides, List is not an instance of SymbolicData while this list is present
-        }
+data ListItem x a = ListItem
+  { tailHash    :: Layout x a
+  , headLayout  :: Layout x a
+  , headPayload :: Payload x a
+  }
+  deriving (Generic1)
+
+deriving instance
+  (Functor (Layout x), Functor (Payload x)) =>
+  Functor (ListItem x)
+
+instance
+    (Representable (Layout x), Representable (Payload x)) =>
+    Distributive (ListItem x) where
+  distribute = distributeRep
+
+instance
+  (Representable (Layout x), Representable (Payload x)) =>
+  Representable (ListItem x)
+
+data List c x = List
+  { lHash    :: c (Layout x)
+  , lSize    :: c Par1
+  , lWitness :: Payloaded (Infinite :.: ListItem x) c
+  }
+  deriving (Generic)
+
+instance HApplicative c => SymbolicData (List c x)
+-- | TODO: Maybe some 'isValid' check for Lists?..
+instance (Symbolic c, SymbolicInput x) => SymbolicInput (List c x)
```

Comment:
Good question! On the one hand, we could check the "knowledge" of the list in `isValid`. That way, we can be sure that the prover has full knowledge of the arguments that they are passing to the computation. On the other hand, I cannot currently construct an example where it would cause a problem.

Reply:
The same applies to the `Hash` types.

Reply:
Well, yeah, we could define
```haskell
isValid (l :: List c x) = foldr (.:) emptyList l == l
isValid (h :: Hash c x) = hash (preimage h) == h
```
But in the end, this would introduce practically the same constraints as the ones user is going to introduce by using the operations on the datatype anyway, so I don't know if this is useful.

Reply:
Ok, let's leave those notes for now and revisit it later.